text/3502-cargo-script.md

```diff
@@ -0,0 +1,1285 @@
+- This makes it harder to resolve conflicts (users can't just checkout the old file and have it re-resolved)
+- A policy is needed when the location is read-only
+  - Fallback to a user-writeable location
+  - Always re-calculate the lockfile
+  - Error
+
+**Configuration 1: Hardcoded**
+
+Unless as a fallback due to a read-only location, the user has no control over
+the lockfile location.
+
+**Configuration 2: Command-line flag**
+
+`cargo generate-lockfile --manifest-path <file>.rs` would be special-cased to
+write the lockfile to the persistent location and otherwise we fallback to a
+no-visible-lockfile solution.
+
+- Passing flags in a `#!` doesn't work cross-platform
+
+**Configuration 3: A new manifest field**
+
+We could add a `workspace.lock` field to control some lockfile location
+behavior, what that is depends on the location on what policies we feel
+comfortable making.  This means we would allow limited access to the
+`[workspace]` table (currently the whole table is verboten).
+
+- Requires manifest design work that is likely specialized to just this feature
+
+**Configuration 4: Exitence Check**
+
+`cargo` can check if the lockfile exists in the agreed-to location and use
+it / update it and otherwise we fallback to a no-visible-lockfile solution.  To
+initially opt-in, a user could place an empty lockfile in that location
+
+**Format 1: Cargo.lock**
+
+We can continue to use the existing `Cargo.lock`.
+
+At this time, just pulling in `clap` and `tokio` includes 51 `[[package]]`
+tables and takes up 419 lines.  This is fine for being an adjacent file but
+might be overwhelming for being embedded.
+We might want to consider ways of reducing redundancy.
+However, at best we can drop the file to 51, 102, or 153 lines (1-3 per package) which can still be overwhelming.
+
+**Format 2: Minimal Versions**
+
+Instead of tracking a distinct lockfile, we can get most of the benefits with
+[`-Zminimal-versions`](https://github.com/rust-lang/cargo/issues/5657).
+
+- Consistent runs across machines without a lockfile
+- More likely to share versions across single-file packages, allowing more
+  reuse within the shared build cache
+- Deviates from how resolution typically happens, surprising people
+- Not all transitive dependencies have valid version requirements
+
+**Format 3: Timestamp**
+
+If we record timestamps with package publishes, we could resolve to a specific
+timestamp for registry packages.
+
+Challenges:
+- Cargo preserves existing resolved versions when dealing with a new or modified
+  dependency or `cargo update -p`.
+- Non-registry dependenciess.
+
+If we want this to be near-lossless, it seems like we'd need
+- An edit log, rather than simple timestamps
+- Regular lockfile entries for non-registry dependencies
+
+See also
+[Cargo time machine (generate lock files based on old registry state) ](https://github.com/rust-lang/cargo/issues/5221)
+
```

Comment:
```suggestion

**Format 4: Minimal lockfile format**

We could simplify the lockfile format to store less information, making it more appealing to embed in the binary. We could store enough information to easily recreate the lockfile, without requiring timestamp-based recreation of the index state. This would primarily include the exact versions of every dependency.
```